### ISSUE TITLE
Why b_IJ is shared between single batch examples.

### ISSUE THREAD

**Issue by pkubik** [naturomics/CapsNet-Tensorflow#21]
Forgive me if I got this wrong but it seems like the `b_IJ` are shared between all examples within a single batch (see [reduce_sum](https://github.com/naturomics/CapsNet-Tensorflow/blob/7832e63e34f3c825ebfba7aff5e3e7dc445753e8/capsLayer.py#L136) and [the shape](https://github.com/naturomics/CapsNet-Tensorflow/blob/7832e63e34f3c825ebfba7aff5e3e7dc445753e8/capsLayer.py#L69)).

I didn't see any mention of the batches in the paper, so I have assumed that there is a separate set of `b_IJ` weights for every batch. Why do you think that it's better to share those variables?

Edit:
I've corrected the statement:
> `b_IJ` are shared between all batches

to:
>  `b_IJ` are shared between all examples within a single batch 

which is was I originally meant.

**Comment by AlexHex7** [comment]
It seems that b_IJ will be init to 0 each batch. Or each time when we invoke __call__ function, the b_IJ will be create to 0. I'm not sure whether my opinion is right, I have not use tensorflow for a long time.

` b_IJ = tf.constant(np.zeros([1, input.shape[1].value, self.num_outputs, 1, 1], dtype=np.float32))`

**Comment by naturomics** [comment]
@AlexHex7 you'r right, b_IJ will be re-init to 0 at each batch.

**Comment by pkubik** [comment]
Oh, sorry. I didn't formulate it correctly. I meant that it (`b_IJ`) is shared between all examples in a single batch. So I'm not sure whether it is correct to share `b_IJ` between different examples in the same batch.

**Comment by pkubik** [comment]
To be more specific what I suggest is to change the [initialization](https://github.com/naturomics/CapsNet-Tensorflow/blob/7832e63e34f3c825ebfba7aff5e3e7dc445753e8/capsLayer.py#L69) of `b_IJ` to:
`b_IJ = tf.constant(np.zeros([cfg.batch_size, input.shape[1].value, self.num_outputs, 1, 1], dtype=np.float32))`
and remove the reduction from the [last line of routing inner loop](https://github.com/naturomics/CapsNet-Tensorflow/blob/7832e63e34f3c825ebfba7aff5e3e7dc445753e8/capsLayer.py#L136):
`b_IJ += u_produce_v`

**Comment by naturomics** [comment]
@pkubik I'm doing a experiment for this problem, please wait for the result of the experiment.

**Comment by naturomics** [comment]
@pkubik Now, I agree with you, though it makes the number of parameters b_IJ be batch_size-related, and experiment shows it doesn't make much difference. [here](https://github.com/XifengGuo/CapsNet-Keras/issues/1) is a related discussion for this problem, It might help us understand why

**Comment by Queequeg92** [comment]
Different sample, different object, different entities, so different b_ij. I'm wondering how much difference it makes@naturomics? Did you make the experiment on minist?

**Comment by naturomics** [comment]
@Queequeg92 Yeah, I did some experiments on mnist. It doesn't seem to make much difference in terms of classification accuracy. So I didn't release the corresponding result. Maybe trying it on the `Fashion-MNIST` that mentioned in issue #20 will see the difference, I will try it soon.
